osquery/tables/system/windows/processes.cpp

```diff
@@ -147,238 +214,334 @@ Status getProcList(std::set<long>& pids) {
-    } else {
-      // Windows stores proc times in 100 nanosecond ticks
-      ULARGE_INTEGER utime;
-      utime.HighPart = userTime.dwHighDateTime;
-      utime.LowPart = userTime.dwLowDateTime;
-      r["user_time"] = BIGINT(utime.QuadPart / 10000);
-      utime.HighPart = kernelTime.dwHighDateTime;
-      utime.LowPart = kernelTime.dwLowDateTime;
-      r["system_time"] = BIGINT(utime.QuadPart / 10000);
-      r["start_time"] = BIGINT(osquery::filetimeToUnixtime(createTime));
-    }
+// On windows 8.1 and later, there's an enum for commandline
+void getProcessCommandLine(HANDLE& proc, std::string& out, unsigned long pid) {
+  if (kNtQueryInformationProcess == nullptr) {
+    VLOG(1) << "Failed to resolve NtQueryInformationProcess with "
+            << GetLastError();
+    return;
   }
 
-  if (context.isAnyColumnUsed({"uid", "gid", "is_elevated_token"})) {
-    /// Get the process UID and GID from its SID
-    HANDLE tok = nullptr;
-    std::vector<char> tokUser(sizeof(TOKEN_USER), 0x0);
-    auto ret = OpenProcessToken(hProcess, TOKEN_READ, &tok);
-    if (ret != 0 && tok != nullptr) {
-      unsigned long tokOwnerBuffLen;
-      ret = GetTokenInformation(tok, TokenUser, nullptr, 0, &tokOwnerBuffLen);
-      if (ret == 0 && GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
-        tokUser.resize(tokOwnerBuffLen);
-        ret = GetTokenInformation(
-            tok, TokenUser, tokUser.data(), tokOwnerBuffLen, &tokOwnerBuffLen);
-      }
+  unsigned long size_out = 0;
+  PROCESS_BASIC_INFORMATION proc_info;
+  SecureZeroMemory(&proc_info, sizeof(PROCESS_BASIC_INFORMATION));
 
-      // Check if the process is using an elevated token
-      auto elevated = FALSE;
-      TOKEN_ELEVATION Elevation;
-      DWORD cbSize = sizeof(TOKEN_ELEVATION);
-      if (GetTokenInformation(
-              tok, TokenElevation, &Elevation, sizeof(Elevation), &cbSize)) {
-        elevated = Elevation.TokenIsElevated;
-      }
+  // See the Process Hacker implementation for more details on the hard coded 60
+  // https://github.com/processhacker/processhacker/blob/master/phnt/include/ntpsapi.h#L160
+  auto ret = kNtQueryInformationProcess(proc, 60, NULL, 0, &size_out);
 
-      r["is_elevated_token"] = elevated ? INTEGER(1) : INTEGER(0);
-    }
-    if (uid != 0 && ret != 0 && !tokUser.empty()) {
-      auto sid = PTOKEN_OWNER(tokUser.data())->Owner;
-      r["uid"] = INTEGER(getUidFromSid(sid));
-      r["gid"] = INTEGER(getGidFromSid(sid));
-    } else {
-      r["uid"] = INTEGER(uid);
-      r["gid"] = INTEGER(gid);
-    }
-    if (tok != nullptr) {
-      CloseHandle(tok);
-      tok = nullptr;
-    }
+  if (ret != STATUS_BUFFER_OVERFLOW && ret != STATUS_BUFFER_TOO_SMALL &&
+      ret != STATUS_INFO_LENGTH_MISMATCH) {
+    VLOG(1) << "NtQueryInformationProcess failed for " << pid << " with "
+            << ret;
+    return;
   }
 
-  if (hProcess != nullptr) {
-    CloseHandle(hProcess);
+  PVOID cmdline = static_cast<PVOID>(malloc(size_out));
```

Comment:
Never freed :droplet:. Track with unique_ptr.

Reply:
God damnit Nick

Reply:
Hrm, surely there's a way to do this with a `unique_ptr` or `shared_ptr`, but I'm failing to find it. I'll make another attempt, but for now I'm just going to `free` it.